subsys/bluetooth/services/hids_c.c

```diff
@@ -0,0 +1,864 @@
+				bt_gatt_dm_attr_chrc_val(gatt_chrc);
+		if (!bt_uuid_cmp(chrc_val->uuid, BT_UUID_HIDS_REPORT)) {
+			++rep_cnt;
+		}
+	}
+	LOG_DBG("%u report(s) found", rep_cnt);
+	if (rep_cnt > UINT8_MAX) {
+		LOG_ERR("Number of records cannot fit into 8 bit identifier.");
+		return -EINVAL;
+	}
+
+	hids_c->rep_info = repptr_array_new(rep_cnt);
+	if (!(hids_c->rep_info)) {
+		LOG_ERR("Cannot create report pointers array.");
+		return -ENOMEM;
+	}
+	hids_c->rep_cnt = (u8_t)rep_cnt;
+
+	/* Process all the records */
+	rep_cnt = 0;
+	gatt_chrc = NULL;
+	while (NULL != (gatt_chrc = bt_gatt_dm_char_next(dm, gatt_chrc))) {
+		struct bt_gatt_chrc *chrc_val =
+				bt_gatt_dm_attr_chrc_val(gatt_chrc);
+		if (!bt_uuid_cmp(chrc_val->uuid, BT_UUID_HIDS_REPORT)) {
+			LOG_DBG("Creating report at index %u, handle: 0x%.4x",
+				rep_cnt, gatt_chrc->handle);
+			ret = rep_new(hids_c,
+				      &(hids_c->rep_info[rep_cnt]),
+				      dm,
+				      gatt_chrc);
+			if (ret) {
+				LOG_ERR("Cannot create report, error: %d", ret);
+				return ret;
+			}
+			++rep_cnt;
+		}
+	}
+
+	LOG_DBG("Report memory allocated, entities used: %u",
+		k_mem_slab_num_used_get(&bt_gatt_hids_c_reports_mem));
+
+	/* Finally - save connection object */
+	hids_c->conn = bt_gatt_dm_conn_get(dm);
+
+	/** @todo This is temporary, as normally there retrieving
+	 * additional data from reports shall start
+	 */
+	hids_c->ready = true;
+	hids_c->ready_cb(hids_c);
+	return 0;
+}
+
+void bt_gatt_hids_c_init(struct bt_gatt_hids_c *hids_c,
+			const struct bt_gatt_hids_c_init_params *params)
+{
+	memset(hids_c, 0, sizeof(*hids_c));
+	hids_c->ready_cb      = params->ready_cb;
+	hids_c->prep_error_cb = params->prep_error_cb;
+	hids_c->pm_update_cb  = params->pm_update_cb;
+	k_sem_init(&hids_c->read_params_sem, 1, 1);
+}
+
+int bt_gatt_hids_c_handles_assign(struct bt_gatt_dm *dm,
+				  struct bt_gatt_hids_c *hids_c)
+{
+	int ret = handles_assign_internal(dm, hids_c);
+
+	if (ret) {
+		bt_gatt_hids_c_release(hids_c);
+	}
+	return ret;
+}
+
+void bt_gatt_hids_c_release(struct bt_gatt_hids_c *hids_c)
+{
+	if (hids_c->rep_info) {
+		u8_t rep_cnt = hids_c->rep_cnt;
+
+		while (rep_cnt) {
+			rep_free(&(hids_c->rep_info[--rep_cnt]));
+		}
+
+		hids_c->rep_cnt = 0;
+		repptr_array_free(hids_c->rep_info);
+		hids_c->rep_info = NULL;
+	}
+	if (hids_c->rep_boot.kbd_inp) {
+		rep_free(&(hids_c->rep_boot.kbd_inp));
+	}
+	if (hids_c->rep_boot.kbd_out) {
+		rep_free(&(hids_c->rep_boot.kbd_out));
+	}
+	if (hids_c->rep_boot.mouse_inp) {
+		rep_free(&(hids_c->rep_boot.mouse_inp));
+	}
+	LOG_DBG("Report memory released, entities used: %u",
+		k_mem_slab_num_used_get(&bt_gatt_hids_c_reports_mem));
+
+	memset(&hids_c->info_val, 0, sizeof(hids_c->info_val));
+	memset(&hids_c->handlers, 0, sizeof(hids_c->handlers));
+	hids_c->map_cb  = NULL;
+	hids_c->ready   = false;
+	hids_c->conn    = NULL;
+	k_sem_give(&hids_c->read_params_sem);
+}
+
+void bt_gatt_hids_c_abort_all(struct bt_gatt_hids_c *hids_c)
+{
+
```

Comment:
If this is left empty on purpose, some comment with explanation would be useful.

Reply:
I have added some code here.